### PR TITLE
support tf2 NAS with non-weight-sharing mode

### PR DESCRIPTION
**Pull request by QuanluZhang**
1. support non-weight-sharing mode for tf2
2. change naive-tf example to run in non-weight-sharing mode
3. add IT for classic nas pytorch

### CODE REVIEW

File: src/sdk/pynni/nni/nas/tensorflow/classic_nas/mutator.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        for mutable in self.mutables:
+            # for now we only generate flattened search space
+            if isinstance(mutable, LayerChoice):
+                key = mutable.key
+                val = mutable.names
+                search_space[key] = {"_type": LAYER_CHOICE, "_value": val}
+            elif isinstance(mutable, InputChoice):
+                key = mutable.key
+                search_space[key] = {"_type": INPUT_CHOICE,
+                                     "_value": {"candidates": mutable.choose_from, #["input_"+str(i) for i in range(mutable.n_candidates)],
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/tensorflow/classic_nas/mutator.py:205]: Please remove useless comments. Or make it explicitly useful.

File: docs/en_US/NAS/NasGuide.md [hunk 2]
```diff
@@ -117,35 +42,13 @@ trainer.train()  # training
...
 
-We support a standalone mode for easy debugging, where you can directly run the trial command without launching an NNI experiment. This is for checking whether your trial code can correctly run. The first candidate(s) are chosen for `LayerChoice` and `InputChoice` in this standalone mode.
+On the other hand, there might be various needs for the search. Some of the needs are embedded in the trainer. For example, some trainers have the assumption that the task is a classification task; some trainers might have a different definition of "epoch" (e.g., an ENAS epoch = some child steps + some controller steps); most trainers do not have support for distributed training: they won't wrap your model with `DataParallel` or `DistributedDataParallel` to do that. So after a few tryouts, if you want to actually use the trainers on your very customized applications, you might need to [customize your trainer](./Advanced.md#extend-the-ability-of-one-shot-trainers).
```
Comment (ultmaster) [inline docs/en_US/NAS/NasGuide.md:49]: Where is `on the one hand`?

File: docs/en_US/NAS/ClassicNas.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+nni.report_final_result(acc)  # report the performance of the chosen architecture
+```
+
+First, instantiate the model. Search space has been defined in this model through `LayerChoice` and `InputChoice`. After that, user should invoke `get_and_apply_next_architecture(model)` to settle down to a specific architecture. This function receives the architecture from tuner (i.e., the classic NAS algorithm) and applies the architecture to `model`. At this point, `model` becomes a specific architecture rather than a search space. Then users are free to train this model just like training a normal PyTorch model. After get the accuracy of this model, users should invoke `nni.report_final_result(acc)` to report the result to the tuner.
+
+At this point, trial code is ready. Then, we can prepare an NNI experiment, i.e., search space file and experiment config file. Different from NNI hyper-parameter tuning, search space file is automatically generated from the trial code by running the command (the detailed usage of this command can be found [here](../Tutorial/Nnictl.md)):
+
+`nnictl ss_gen --trial_command="the command for running your trial code"`
```
Comment (ultmaster) [inline docs/en_US/NAS/ClassicNas.md:23]: I think we should use `-t` because it's simpler.

File: docs/en_US/NAS/ClassicNas.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Classic NAS Algorithms
+
+In classic NAS algorithms, each architecture is trained as a trial and the NAS algorithm acts as a tuner. Thus, this training mode naturally fits within the NNI hyper-parameter tuning framework, where Tuner generates new architecture for the next trial and trials run in the training service.
```
Comment (ultmaster) [inline docs/en_US/NAS/ClassicNas.md:3]: , where tuners generate ...

File: examples/nas/classic_nas-tf/train.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    
+    get_and_apply_next_architecture(net)
+
+    optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+
+    train(net, dataset_train, optimizer)
+
+    acc = test(net, dataset_test)
+
+    nni.report_final_result(acc.numpy())
```
Comment (ultmaster) [inline examples/nas/classic_nas-tf/train.py:125]: The reported result is a numpy item?

File: examples/nas/classic_nas-tf/train.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            epoch_loss_avg.update_state(loss_value)
+            epoch_accuracy.update_state(y, net(x, training=True))
+
+        train_loss_results.append(epoch_loss_avg.result())
+        train_accuracy_results.append(epoch_accuracy.result())
+
+        if epoch % 1 == 0:
+            print("Epoch {:03d}: Loss: {:.3f}, Accuracy: {:.3%}".format(epoch,
+                                                                epoch_loss_avg.result(),
+                                                                epoch_accuracy.result()))
```
Comment (ultmaster) [inline examples/nas/classic_nas-tf/train.py:91]: Should we report intermediate results here?

File: examples/nas/classic_nas-tf/train.py (hunk 5 above)
Comment (QuanluZhang) [reply examples/nas/classic_nas-tf/train.py:125]: yes, it is a numpy scalar

File: src/sdk/pynni/nni/nas/tensorflow/classic_nas/mutator.py (hunk 1 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/tensorflow/classic_nas/mutator.py:205]: sure, forgot to remove

File: examples/nas/classic_nas-tf/train.py (hunk 6 above)
Comment (QuanluZhang) [reply examples/nas/classic_nas-tf/train.py:91]: no need to report intermediate result, because this example does not use it. do you think this example should use assessor?

File: docs/en_US/NAS/NasGuide.md (hunk 2 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md:49]: updated this part

File: docs/en_US/NAS/ClassicNas.md (hunk 3 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/ClassicNas.md:23]: I think use `--trial_command` is clear

File: docs/en_US/NAS/ClassicNas.md (hunk 4 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/ClassicNas.md:3]: I prefer to keep it. or change it to "where a tuner generates"

File: examples/nas/classic_nas-tf/train.py (hunk 6 above)
Comment (ultmaster) [reply examples/nas/classic_nas-tf/train.py:91]: I think it's for demo purposes. It's fine without it.

File: test/nni_test/nnitest/run_tests.py [hunk 7]
```diff
@@ -51,6 +65,12 @@ def prepare_config_file(test_case_config, it_config, args):
     if sys.platform == 'win32' and args.ts == 'local':
         test_yml_config['trial']['command'] = test_yml_config['trial']['command'].replace('python3', 'python')
 
+    # generate search space file for classic nas
+    if test_case_config.get('doSsgen') is not None:
```
Comment (ultmaster) [inline test/nni_test/nnitest/run_tests.py:69]: doSsGen?

File: docs/en_US/nas.rst [hunk 8]
```diff
@@ -18,15 +18,9 @@ For details, please refer to the following tutorials:
...
-    ENAS <NAS/ENAS>
-    DARTS <NAS/DARTS>
-    P-DARTS <NAS/PDARTS>
-    SPOS <NAS/SPOS>
-    CDARTS <NAS/CDARTS>
-    ProxylessNAS <NAS/Proxylessnas>
-    TextNAS <NAS/TextNAS>
+    Write A Search Space <NAS/QuickStart>
+    Classic NAS <NAS/ClassicNas>
+    One-shot NAS <NAS/one_shot_nas>
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply docs/en_US/nas.rst:23]: what do you mean?

File: test/nni_test/nnitest/run_tests.py [hunk 9]
```diff
@@ -39,6 +39,20 @@ def update_training_service_config(config, training_service):
     deep_update(config, it_ts_config['all'])
     deep_update(config, it_ts_config[training_service])
 
+def nnictl_generate_search_space(test_yml_config, test_case_config, args):
+    trial_command = test_yml_config['trial']['command']
+    code_dir = args.nni_source_dir + test_case_config['ssgenCodeDir']
+    ss_file_path = args.nni_source_dir + test_case_config['ssFilePath']
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply test/nni_test/nnitest/run_tests.py:45]: good suggestion, I just followed the previous code

File: test/config/integration_tests.yml [hunk 10]
```diff
@@ -72,6 +72,14 @@ testCases:
...
   configFile: test/config/examples/mnist-nested-search-space.yml
 
+- name: classic-nas-pytorch
+  configFile: test/config/examples/classic-nas-pytorch.yml
+  # generate search space through nnictl ss_gen
+  doSsgen: True
+  ssgenCodeDir: examples/nas/classic_nas
+  # this file is automatically generated by nnictl ss_gen
+  ssFilePath: test/config/examples/nni-nas-search-space.json
+
```
Comment (chicm-ms) [inline test/config/integration_tests.yml:83]: Maybe we can split this test cases into 2 cases, then we will do not need handle the special key 'doSsgen' etc only for this case.
1. 
```
name: nas-gen-ss
launchCommand: nnictl ss_gen --trial_command="python3 mnist.py --epochs 1" --trial_dir=examples/nas/classic_nas --file=test/config/examples/nni-nas-search-space.json
stopCommand:
experimentStatusCheck: False
```

2.
```
name: classic-nas-pytorch
configFile: test/config/examples/classic-nas-pytorch.yml
```

File: test/config/integration_tests.yml [hunk 11]
```diff
@@ -72,6 +72,14 @@ testCases:
 - name: nested-ss
   configFile: test/config/examples/mnist-nested-search-space.yml
 
+- name: classic-nas-pytorch
```
Comment (chicm-ms) [inline test/config/integration_tests.yml:75]: Is this a tf2 test case?  if yes, need to be moved into test/config/integration_tests_tf2.yml

Comment (QuanluZhang) [reply test/config/integration_tests.yml:75]: this one can be tested with tf 1.x, I will add another one in `test/config/integration_tests_tf2.yml`

File: test/config/integration_tests.yml (hunk 10 above)
Comment (QuanluZhang) [reply test/config/integration_tests.yml:83]: good point

Comment (QuanluZhang) [reply test/config/integration_tests.yml:83]: by the way, are the two cases guaranteed to be tested sequentially one after the other?

Comment (chicm-ms) [reply test/config/integration_tests.yml:83]: Yes, the order is guaranteed.   You can run the 2 cases on local system to check whether the case is OK:
```
python3 nni_test/nnitest/run_test.py --config config/integration_tests.yml --case classic-nas
```

File: docs/en_US/NAS/NasGuide.md [hunk 12]
```diff
@@ -117,35 +42,13 @@ trainer.train()  # training
 trainer.export(file="model_dir/final_architecture.json")  # export the final architecture to file
```
Comment (chicm-ms) [inline docs/en_US/NAS/NasGuide.md:42]: `reward_accuracy` is not defined.   should be `top1_accuracy`

File: test/config/integration_tests.yml (hunk 10 above)
Comment (QuanluZhang) [reply test/config/integration_tests.yml:83]: seems stopCommand cannot be the combine of two commands....

Comment (chicm-ms) [reply test/config/integration_tests.yml:85]: yes, it seems the combination of two commands not working for `stopCommand` due to the `shlex.split`,  
it seems it can be fixed by replacing `proc = subprocess.run(shlex.split(launch_command))` with:
```
proc = subprocess.run(launch_command, shell=True)
```

File: test/nni_test/nnitest/run_tests.py [hunk 13]
```diff
@@ -75,7 +75,9 @@ def run_test_case(test_case_config, it_config, args):
         stop_command = get_command(test_case_config, 'stopCommand')
         print('Stop command:', stop_command, flush=True)
         if stop_command:
-            subprocess.run(shlex.split(stop_command))
+            for command in stop_command.split('&'):
+                print('Command:', command, flush=True)
```
Comment (chicm-ms) [inline test/nni_test/nnitest/run_tests.py:79]: better to fix it like this to support multiple commands:
```
subprocess.run(launch_command, shell=True)
```

File: docs/en_US/NAS/NasGuide.md (hunk 12 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md:42]: @ultmaster could you answer this question?

Comment (ultmaster) [reply docs/en_US/NAS/NasGuide.md:42]: In the ENAS exmaple, `reward_accuracy` is the same as `top1_accuracy`, although they can be different (e.g., `reward_fn` can be a joint function of accuracy and latency). For this documentation, I'm okay with changing into `top1_accuracy` to avoid possible confusion.

File: docs/en_US/nas.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (ultmaster) [reply docs/en_US/nas.rst:23]: `one_shot_nas` -> `OneShotNas`.

File: test/nni_test/nnitest/run_tests.py [hunk 14]
```diff
@@ -76,6 +76,10 @@ def run_test_case(test_case_config, it_config, args):
         print('Stop command:', stop_command, flush=True)
         if stop_command:
             subprocess.run(shlex.split(stop_command))
+        exit_command = get_command(test_case_config, 'onExitCommand')
+        print('Exit command:', exit_command, flush=True)
+        if exit_command:
+            subprocess.run(shlex.split(exit_command))
```
Comment (ultmaster) [inline test/nni_test/nnitest/run_tests.py:82]: Add `check=True` as subprocess doesn't throw an exception by default.

File: docs/en_US/nas.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply docs/en_US/nas.rst:23]: this is .rst file, it has different naming convention from .md file

File: docs/en_US/NAS/NasGuide.md (hunk 12 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md:42]: sure, will update to `top1_accuracy`

File: test/nni_test/nnitest/run_tests.py (hunk 14 above)
Comment (QuanluZhang) [reply test/nni_test/nnitest/run_tests.py:82]: good point, thanks.
